Barebones template function for code generation

hugomg opened the pull request:
We replaced all uses of string.format for generating C code with a little function that used named placehoders instead of positional placeholders.

There are also some very minor code cleanups (whitespace fixes, C literal creation, and C identifier creation)

----

Bikesheding moment: I don't like the "output" name for the function because it's ambiguous about being a noun or verb and because it isn't specific to what the function does. But I couldn't think of something better :/

hishamhm committed:
Coder: add output template function

hishamhm committed:
Coder: convert uses of string.format() to output()

Commit:
Clean some end-of-line whitespace

My editor did this. I tried to put it in a separate commit at least.

Commit:
Get rid of side effects in the template function

By not setting the metatable for the substitution parameter we keep the
template function side-effect free. It should also behave better if the
substitution table already has a metatable set (though I can't think of
why we would want to do that).

Commit:
Finish replacing string.format -> output

Commit:
Separate functions for creating C literals

Commit:
Stop using template function to create identifiers

It was a bit confusing that $VAR_blah and blah_$VAR didn't do the same
thing. So I changed all the cases to do the string concatenation outside
the template.

hishamhm commented on titan-compiler/coder.lua:
```diff
@@ -238,81 +338,101 @@ local function codefor(ctx, node)
+                    cstep = output("$CVAR = l_castU2S(l_castS2U($CVAR) - $NEGILIT)", subs)
                 else
-                    cstep = node.decl._cvar .. " -= " .. tostring(ilit)
+                    cstep = output("$CVAR -= $ILIT", subs)
```
Re-reading the code I just spotted a bug (that I wrote myself)... ILIT is not defined in subs at this point. The fact that this wasn't caught by Busted means we don't have a test covering this line!

hishamhm replied:
Pushed a fix for this, and tests

hishamhm commented on titan-compiler/coder.lua:
```diff
@@ -550,46 +722,75 @@ local function codetable(ctx, node, target)
     local cinit, ctmp, tmpname, tmpslot
     if target then
+        -- TODO: double check this code, it wan't covered by tests
+        -- and wan't passing anything to the second $TMPNAME placeholder
```
Good catch! The new template function will avoid these things in the future :)

hishamhm commented on titan-compiler/coder.lua:
```diff
@@ -625,16 +826,25 @@ local function codebinaryop(ctx, node, iscondition)
+                $CTMP
+                $TMPNAME = $LCODE;
+                if($TMPNAME) {
+                  $RSTATS
```
(nitpick: the C code templates above were intented with 4 spaces)

hishamhm commented:
> Bikesheding moment: I don't like the "output" name for the function because it's ambiguous about being a noun or verb and because it isn't specific to what the function does. But I couldn't think of something better :/

@hugomg A valid concern, though it didn't seem ambiguous to me because I try to make all function names to be verbs. I was out of ideas for alternatives myself, so I outsourced the brainstorming: 😄 

https://twitter.com/hisham_hm/status/928977889265143809

hishamhm committed:
Coder: fix handling of float literals in 'for'

hishamhm committed:
Tests: float values in 'for'

hishamhm committed:
Coder: rename output to render

The old name was considered ambiguous by @hugomg.
Thanks @kikito for the suggestion!

hishamhm committed:
Coder: indentation tweaks
